### DISCUSSION TITLE
[Typescirpt] How to compile `export default` to `module.exports` and even in the type declaration?

### DISCUSSION THREAD

**Discussion by MTtankkeo** [community/community#137001] [Programming Help] [answered]
### Body

Hello,

I'm currently working on an npm package for a Webpack plugin. I'm using TypeScript and compiling to CommonJS JavaScript without using a separate bundler.

The issue I'm encountering is that when I export a module using export default, I have to import it with require("webpack-plugin").default, which seems to be an extra step. This doesn't seem like the intended behavior of default exports.

To address this, I tried changing the export default in my TypeScript code to module.exports. This fixed the issue for the runtime code, but the problem persisted in the type declaration files.

Is there a known solution or workaround for this issue? Any help would be greatly appreciated.

Thanks.

## tsconfig.json
```json
{
    "compilerOptions": {
        "lib": ["ES2023", "ES2023.Collection", "DOM"],
        "module": "CommonJS",
        "target": "ES2022",
        "outDir": "./dist/",
        "moduleResolution": "Node",
        "skipLibCheck": true,
        "declaration": true,
        "declarationDir": "./dist/types/"
    },
    "include": ["./src/**/*"],
    "exclude": ["./node_modules"]
}
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by lqduy** [comment]
# Use `esModuleInterop` or `allowSyntheticDefaultImports` in `tsconfig.json`

Adding `"esModuleInterop": true` or `"allowSyntheticDefaultImports": true` to your `tsconfig.json` makes TypeScript handle default imports more seamlessly when compiling to CommonJS. It allows you to use both `require('module')` and `import module from 'module'` interchangeably.

```json
{
  "compilerOptions": {
    "lib": ["ES2023", "ES2023.Collection", "DOM"],
    "module": "CommonJS",
    "target": "ES2022",
    "outDir": "./dist/",
    "moduleResolution": "Node",
    "skipLibCheck": true,
    "declaration": true,
    "declarationDir": "./dist/types/",
    "esModuleInterop": true,  // Add this option
    "allowSyntheticDefaultImports": true  // Optionally add this
  }
}
```

This option should fix the `.default` issue when using `require()` and also preserve the proper type declarations.

**Comment by lqduy** [comment, marked as answer]
# Use `module.exports` for Runtime and `export =` for Types

Another approach is to use `module.exports` in your JavaScript output for the runtime but leverage TypeScript's `export =` syntax for type declarations.

In your `.ts` code:

```ts
class WebpackPlugin {
  // Plugin implementation
}

export = WebpackPlugin;
```

This makes `require('webpack-plugin')` work as expected and aligns well with CommonJS usage. You can also use `import WebpackPlugin = require('webpack-plugin');` in other TypeScript files.

**Comment by lqduy** [comment]
# Hybrid Approach with `module.exports` and declare class

Alternatively, you can explicitly set `module.exports` in your TypeScript code while declaring the types separately:

```ts
class WebpackPlugin {
  // Your class implementation
}

module.exports = WebpackPlugin;
```

And in a separate `.d.ts` file or within the same file:

```ts
declare class WebpackPlugin {
  // Declare the plugin's type information here
}

export = WebpackPlugin;
```
